daemon/algod/api/server/v2/dryrun.go

```diff
@@ -0,0 +1,295 @@
+	Round uint64 `codec:"round,omitempty"`
+}
+
+type dryrunDebugReceiver struct {
+	history       []logic.DebugState
+	scratchActive []bool
+}
+
+func (ddr *dryrunDebugReceiver) updateScratch() {
+	any := false
+	maxActive := -1
+	lasti := len(ddr.history) - 1
+
+	for i, sv := range ddr.history[lasti].Scratch {
+		if sv.Type != "u" || sv.Uint != 0 {
+			any = true
+			maxActive = i
+		}
+	}
+	if any {
+		if ddr.scratchActive == nil {
+			ddr.scratchActive = make([]bool, maxActive+1, 256)
+		}
+		for i := len(ddr.scratchActive); i <= maxActive; i++ {
+			sv := ddr.history[lasti].Scratch[i]
+			active := sv.Type != "u" || sv.Uint != 0
+			ddr.scratchActive = append(ddr.scratchActive, active)
+		}
+	} else {
+		if ddr.scratchActive != nil {
+			ddr.history[lasti].Scratch = ddr.history[lasti].Scratch[:len(ddr.scratchActive)]
+		} else {
+			ddr.history[lasti].Scratch = nil
+			return
+		}
+	}
+	scratchlen := maxActive + 1
+	if len(ddr.scratchActive) > scratchlen {
+		scratchlen = len(ddr.scratchActive)
+	}
+	ddr.history[lasti].Scratch = ddr.history[lasti].Scratch[:scratchlen]
+	for i := range ddr.history[lasti].Scratch {
+		if !ddr.scratchActive[i] {
+			ddr.history[lasti].Scratch[i].Type = ""
+		}
+	}
+}
+
+// Register is fired on program creation (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Register(state *logic.DebugState) error {
+	ddr.history = append(ddr.history, *state)
+	ddr.updateScratch()
+	return nil
+}
+
+// Update is fired on every step (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Update(state *logic.DebugState) error {
+	// see go-algorand/data/transactions/logic/debugger.go refreshDebugState() for which fields are updated
+	ds := logic.DebugState{
+		PC:      state.PC,
+		Line:    state.Line,
+		Error:   state.Error,
+		Stack:   state.Stack,
+		Scratch: state.Scratch,
+	}
+	ddr.history = append(ddr.history, ds)
+	ddr.updateScratch()
+	return nil
+}
+
+// Complete is called when the program exits (DebuggerHook interface)
+func (ddr *dryrunDebugReceiver) Complete(state *logic.DebugState) error {
+	return ddr.Update(state)
+}
+
+// LedgerForLogic
+type dryrunLedger struct {
+	dr *DryrunRequest
+}
+
+// LedgerForLogic
+func (dl *dryrunLedger) Balance(addr basics.Address) (basics.MicroAlgos, error) {
+	return basics.MicroAlgos{Raw: 0}, nil
+}
+func (dl *dryrunLedger) Round() basics.Round {
+	return basics.Round(dl.dr.Round)
+}
+func (dl *dryrunLedger) AppGlobalState(appIdx basics.AppIndex) (basics.TealKeyValue, error) {
+	for _, app := range dl.dr.Apps {
+		if app.AppIndex == uint64(appIdx) {
+			return app.Params.GlobalState, nil
+		}
+	}
+	return nil, nil
+}
+func (dl *dryrunLedger) AppLocalState(addr basics.Address, appIdx basics.AppIndex) (basics.TealKeyValue, error) {
+	for _, st := range dl.dr.AccountAppStates {
+		if appIdx == st.AppIndex && addr == st.Account {
+			return st.State.KeyValue, nil
```

Comment:
this needs to allocate a `TealKeyValue` if `st.State.KeyValue` is `nil` (`ledger/applications.go`)